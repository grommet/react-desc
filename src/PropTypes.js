// array function should not be used here so that we can the local this
const addPropTypeDocumentationField = fieldName =>
  function addFieldToReactDesc(value) {
    if (!this.reactDesc) {
      this.reactDesc = {};
    }
    this.reactDesc[fieldName] = value;
    return this;
  };

const documentedPropType = {
  defaultValue: addPropTypeDocumentationField('defaultValue'),
  description: addPropTypeDocumentationField('description'),
  deprecated: addPropTypeDocumentationField('deprecated'),
  format: addPropTypeDocumentationField('format')
};

const createPropType = type => {
  const propTypeObj = {
    type,
    ...documentedPropType
  };
  Object.defineProperty(propTypeObj, 'isRequired', {
    get: function getRequired() {
      if (!this.reactDesc) {
        this.reactDesc = {};
      }
      this.reactDesc.required = true;
      return this;
    },
    enumerable: true,
    configurable: true
  });
  return propTypeObj;
};

const createPropTypeWithArgs = type => args => {
  const propTypeObj = {
    args,
    type,
    ...documentedPropType
  };
  Object.defineProperty(propTypeObj, 'isRequired', {
    get: function getRequired() {
      if (!this.reactDesc) {
        this.reactDesc = {};
      }
      this.reactDesc.required = true;
      return this;
    },
    enumerable: true,
    configurable: true
  });
  return propTypeObj;
};

const PropTypes = {
  custom: callback => {
    const target = callback.bind(null);
    target.type = 'func';
    Object.keys(documentedPropType).forEach(fieldName => {
      target[fieldName] = documentedPropType[fieldName];
    });
    return target;
  }
};

function definePropType(type) {
  Object.defineProperty(PropTypes, type, {
    get: function getPropType() {
      return createPropType(type);
    },
    enumerable: true,
    configurable: true
  });
}

function definePropTypeWithArgs(type) {
  Object.defineProperty(PropTypes, type, {
    get: function getPropType() {
      return createPropTypeWithArgs(type);
    },
    enumerable: true,
    configurable: true
  });
}

definePropType('any');
definePropType('array');
definePropType('bool');
definePropType('element');
definePropType('func');
definePropType('node');
definePropType('number');
definePropType('object');
definePropType('symbol');
definePropType('string');

definePropTypeWithArgs('arrayOf');
definePropTypeWithArgs('instanceOf');
definePropTypeWithArgs('objectOf');
definePropTypeWithArgs('oneOfType');
definePropTypeWithArgs('oneOf');
definePropTypeWithArgs('shape');

export default PropTypes;
