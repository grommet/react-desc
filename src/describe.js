import PropTypes from 'prop-types';
import descToJSON from './descToJSON';
import descToMarkdown from './descToMarkdown';
import descToTypescript from './descToTypescript';

const convertArray = array =>
  array.map(type => convertPropType(type));

const convertShape = (shape) => {
  const result = {};
  Object.keys(shape).forEach((key) => {
    result[key] = convertPropType(shape[key]);
  });
  return result;
};

const convertPropType = (propType) => {
  let result;
  if (propType && propType.type) {
    if (!PropTypes[propType.type]) {
      throw new Error(`react-desc: unknown type ${propType.type}`);
    }
    if (propType.args) {
      if (propType.type === 'oneOfType' || propType.type === 'arrayOf') {
        if (Array.isArray(propType.args)) {
          result = PropTypes[propType.type](convertArray(propType.args));
        } else {
          result = PropTypes[propType.type](convertPropType(propType.args));
        }
      } else if (propType.type === 'shape') {
        result = PropTypes[propType.type](convertShape(propType.args));
      } else {
        result = PropTypes[propType.type](propType.args);
      }
    } else {
      result = PropTypes[propType.type];
    }
  } else {
    result = propType;
  }
  return result;
};

export default function describe(ComponentInstance) {
  if (!ComponentInstance) {
    throw new Error('react-desc: component is required');
  }

  const documentation = {
    propTypes: {},
  };

  const DocumentedComponent = ComponentInstance;

  const addDocumentationProp = propName => (value) => {
    documentation[propName] = value;
    return DocumentedComponent;
  };

  DocumentedComponent.availableAt = addDocumentationProp('availableAt');
  DocumentedComponent.description = addDocumentationProp('description');
  DocumentedComponent.details = addDocumentationProp('details');
  DocumentedComponent.deprecated = addDocumentationProp('deprecated');
  DocumentedComponent.usage = addDocumentationProp('usage');
  DocumentedComponent.intrinsicElement = addDocumentationProp('intrinsicElement');

  DocumentedComponent.toJSON = descToJSON.bind(null, ComponentInstance, documentation);
  DocumentedComponent.toTypescript = descToTypescript.bind(null, ComponentInstance, documentation);
  DocumentedComponent.toMarkdown = descToMarkdown.bind(null, ComponentInstance, documentation);

  const externalDocumentation = {
    toJSON: DocumentedComponent.toJSON,
    toMarkdown: DocumentedComponent.toMarkdown,
    toTypescript: DocumentedComponent.toTypescript,
  };

  Object.defineProperty(
    DocumentedComponent,
    'propTypes', {
      get: () => DocumentedComponent.propTypesValue,
      set: (value) => {
        if (!DocumentedComponent.propTypesValue) {
          DocumentedComponent.propTypesValue = {};
        }
        Object.keys(value).forEach((name) => {
          let propType = value[name];
          if (propType.type) {
            documentation.propTypes[name] = propType;
            propType = convertPropType(propType);
            if (value[name].reactDesc.required) {
              propType = propType.isRequired;
            }
          }
          DocumentedComponent.propTypesValue[name] = propType;
          return propType;
        });
      },
      enumerable: true,
      configurable: true,
    },
  );

  Object.defineProperty(
    DocumentedComponent,
    'describedPropTypes', {
      get: () => documentation.propTypes,
      enumerable: true,
      configurable: true,
    },
  );

  Object.defineProperty(
    DocumentedComponent,
    'documentation', {
      get: () => externalDocumentation,
      enumerable: true,
      configurable: true,
    },
  );

  return DocumentedComponent;
}
