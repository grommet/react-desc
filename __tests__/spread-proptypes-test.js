import React from 'react';

import describe from '../src/describe';
import PropTypes from '../src/PropTypes';

class FakeComponent {}

const baseProps = {
  color: PropTypes.string.description('color (string)'),
  fontSize: PropTypes.number.description('fontSize (number)'),
  one: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.instanceOf(FakeComponent),
  ]).description('one (oneOfType)'),
  deprecatedProp: PropTypes.string
    .description('deprecated (string)')
    .deprecated('do not use'),
};

const additionalProps = {
  arr: PropTypes.arrayOf(PropTypes.number).description('arr (array of number)'),
  fun: PropTypes.func
    .description('fun (function)')
    .deprecated('a deprecated function'),
};

const BaseComponent = () => <div>base</div>;
const DocumentedBaseComponent = describe(BaseComponent).description(
  'the base component',
);
DocumentedBaseComponent.propTypes = baseProps;

const ExtendedComponent = () => <div>extended</div>;
const DocumentedExtendedComponent = describe(ExtendedComponent).description(
  'this component spreads the base component prop types',
);
DocumentedExtendedComponent.propTypes = {
  ...BaseComponent.describedPropTypes,
  ...additionalProps,
};

it('returns JSON with all props', () => {
  expect(DocumentedExtendedComponent.toJSON()).toMatchSnapshot();
});

it('returns Markdown with all props', () => {
  expect(DocumentedExtendedComponent.toMarkdown()).toMatchSnapshot();
});

it('returns Typescript with all props', () => {
  expect(DocumentedExtendedComponent.toTypescript()).toMatchSnapshot();
});
