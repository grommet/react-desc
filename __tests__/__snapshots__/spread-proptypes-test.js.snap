// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns JSON with all props 1`] = `
Object {
  "description": "this component spreads the base component prop types",
  "name": "ExtendedComponent",
  "properties": Array [
    Object {
      "description": "color (string)",
      "format": "string",
      "name": "color",
    },
    Object {
      "description": "fontSize (number)",
      "format": "number",
      "name": "fontSize",
    },
    Object {
      "description": "one (oneOfType)",
      "format": "string
number
new FakeComponent(...)",
      "name": "one",
    },
    Object {
      "deprecated": "do not use",
      "description": "deprecated (string)",
      "format": "string",
      "name": "deprecatedProp",
    },
    Object {
      "description": "arr (array of number)",
      "format": "[number]",
      "name": "arr",
    },
    Object {
      "deprecated": "a deprecated function",
      "description": "fun (function)",
      "format": "function",
      "name": "fun",
    },
  ],
}
`;

exports[`returns Markdown with all props 1`] = `
"## ExtendedComponent
this component spreads the base component prop types


## Properties

**color**

color (string)

\`\`\`
string
\`\`\`

**fontSize**

fontSize (number)

\`\`\`
number
\`\`\`

**one**

one (oneOfType)

\`\`\`
string
number
new FakeComponent(...)
\`\`\`

**~~deprecatedProp~~** (do not use)

deprecated (string)

\`\`\`
string
\`\`\`

**arr**

arr (array of number)

\`\`\`
[number]
\`\`\`

**~~fun~~** (a deprecated function)

fun (function)

\`\`\`
function
\`\`\`
  "
`;

exports[`returns Typescript with all props 1`] = `
Object {
  "description": "this component spreads the base component prop types",
  "name": "ExtendedComponent",
  "properties": Array [
    Object {
      "description": "color (string)",
      "format": "string",
      "name": "color",
    },
    Object {
      "description": "fontSize (number)",
      "format": "number",
      "name": "fontSize",
    },
    Object {
      "description": "one (oneOfType)",
      "format": "string | number | any",
      "name": "one",
    },
    Object {
      "deprecated": "do not use",
      "description": "deprecated (string)",
      "format": "string",
      "name": "deprecatedProp",
    },
    Object {
      "description": "arr (array of number)",
      "format": "number[]",
      "name": "arr",
    },
    Object {
      "deprecated": "a deprecated function",
      "description": "fun (function)",
      "format": "((...args: any[]) => any)",
      "name": "fun",
    },
  ],
}
`;
